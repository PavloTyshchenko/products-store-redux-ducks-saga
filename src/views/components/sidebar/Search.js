import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Form } from 'react-bootstrap';

import * as productDuck from '../../../ducks/products';

export const Search = (props) => {

    const [text, setText] = useState('');

    const {
        history,
        match,
        category,
        searchProducts
    } = props;

    // add same search term in field if there is param in url
    useEffect(() => {
        if (match.params.term)
            setText(match.params.term);
        // eslint-disable-next-line
    }, []);

    const onChange = (e) => {
        setText(e.target.value);
    };

    const onSubmit = (e) => {
        e.preventDefault();

        if (text !== '') {
            history.push(`/products/${category}/${text}`);
            searchProducts(category, text);
        }
    };

    return (
        <Form onSubmit={onSubmit} data-test="searchComponent">
            <Form.Group>
                <Form.Control data-test="search-input"
                    type="text"
                    onChange={onChange}
                    value={text}
                    placeholder="Search Products..." />
            </Form.Group>
        </Form>
    );
};

const mapStateToProps = (state) => {
    return {
        category: state.products.category,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        searchProducts: (category, search_term) => {
            dispatch(productDuck.searchProducts(category, search_term));
        }
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Search));
