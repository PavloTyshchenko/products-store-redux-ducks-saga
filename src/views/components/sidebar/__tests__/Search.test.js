import React from 'react';
import { shallow } from 'enzyme';

import { Search } from '../Search';

const defaultProps = {
    history: '',
    match: '',
    category: 'All',
    searchProducts: jest.fn()
};

const setup = (propsOption) => {
    const props = Object.assign(defaultProps, propsOption);
    const wrapper = shallow(<Search {...props} />);
    return {
        wrapper,
        props
    };
};

describe('Search Component', () => {

    test('Should render without errors', () => {
        const { wrapper } = setup();
        const component = wrapper.find(`[data-test='searchComponent']`);
        expect(component.length).toBe(1);
    });

    test('Should match snapshot', () => {
        const component = shallow(<Search />);
        expect(component.debug()).toMatchSnapshot();
    });
});