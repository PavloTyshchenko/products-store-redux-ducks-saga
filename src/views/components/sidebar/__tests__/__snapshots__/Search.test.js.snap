// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Should match snapshot 1`] = `
"<Form onSubmit={[Function: onSubmit]} data-test=\\"searchComponent\\" inline={false}>
  <FormGroup>
    <FormControl data-test=\\"search-input\\" type=\\"text\\" onChange={[Function: onChange]} value=\\"\\" placeholder=\\"Search Products...\\" />
  </FormGroup>
</Form>"
`;
